aws/aws.go

```diff
@@ -454,9 +455,37 @@ func deleteVolumes(ec2Clients map[string]ec2Client, volumes []*types.Disk) []err
 				if ctx.DryRun {
 					log.Infof("[AWS] Dry-run set, volume is not deleted: %s:%s, region: %s", vol.Name, vol.ID, region)
 				} else {
-					log.Infof("[AWS] Delete volume: %s:%s", vol.Name, vol.ID)
-					if _, err := ec2Client.DeleteVolume(&ec2.DeleteVolumeInput{VolumeId: &vol.ID}); err != nil {
-						errChan <- err
+					log.Infof("[AWS] Initiate delete volume: %s:%s", vol.Name, vol.ID)
+					var detachError error
+					if vol.State == types.InUse {
+						log.Infof("[AWS] Volume %s:%s is in-use, trying to detach", vol.Name, vol.ID)
+						if _, detachError = ec2Client.DetachVolume(&ec2.DetachVolumeInput{VolumeId: &vol.ID}); detachError == nil {
+							log.Infof("[AWS] Volume %s:%s detach succeeded, waiting for 'available' state...", vol.Name, vol.ID)
+							//Polling state max 10 times with 1 sec interval
+							var counter int = 0
+							d, e := getDisk(ec2Client, vol.ID)
+							for d != nil && d.State != types.Unused && e == nil && counter < 10 {
+								time.Sleep(1 * time.Second)
+								d, e = getDisk(ec2Client, vol.ID)
+								counter++
+							}
+							if d != nil && d.State == types.Unused {
+								log.Infof("[AWS] Volume %s:%s is detached so it can be deleted.", vol.Name, vol.ID)
+							} else {
+								detachError = errors.New("Failed to verify un-used state.")
+							}
+						}
+					} else {
+						log.Infof("[AWS] Volume %s:%s is not in-use so it can be deleted.", vol.Name, vol.ID)
+					}
+
+					if detachError == nil {
+						log.Infof("[AWS] Delete volume: %s:%s", vol.Name, vol.ID)
+						if _, err := ec2Client.DeleteVolume(&ec2.DeleteVolumeInput{VolumeId: &vol.ID}); err != nil {
+							errChan <- err
+						}
+					} else {
```

Comment:
Early return/continue instead of else please.

Reply:
What's your point on early returning here? For me this format is more readable as the else should happen rarely.

Reply:
It is more goish for two reasons; usually we handle error first and main function goes to the left as much as possible.

Reply:
Thanks for the explanation